src/core/server/saved_objects/permission_control/acl.ts

```diff
@@ -0,0 +1,268 @@
+    if ((!users && !groups) || !permissionTypes) {
+      return this;
+    }
+    if (!this.permissions) {
+      this.permissions = {};
+    }
+
+    for (const permissionType of permissionTypes) {
+      this.permissions[permissionType] = addPermissionToPrincipals(
+        this.permissions[permissionType],
+        users,
+        groups
+      );
+    }
+
+    return this;
+  }
+
+  // permissions object build funciton, remove specific permission of specific users or groups from the object
+  public removePermission(permissionTypes: string[], users?: string[], groups?: string[]) {
+    if ((!users && !groups) || !permissionTypes) {
+      return this;
+    }
+    if (!this.permissions) {
+      this.permissions = {};
+    }
+
+    for (const permissionType of permissionTypes) {
+      this.permissions[permissionType] = deletePermissionFromPrincipals(
+        this.permissions[permissionType],
+        users,
+        groups
+      );
+    }
+
+    return this;
+  }
+
+  // return the permissions object
+  public getPermissions() {
+    const permissions = this.permissions;
+    // reset permissions
+    this.permissions = {};
+
+    return permissions;
+  }
+
+  /*
+    generate query DSL by the specific conditions, used for fetching saved objects from the saved objects index
+    */
+  public genereateGetPermittedSavedObjectsQueryDSL(
+    permissionType: string,
+    savedObjectType?: string | string[],
+    user?: string,
+    group?: string
+  ) {
+    if ((!user && !group) || !permissionType) {
+      return {
+        query: {
+          match_none: {},
+        },
+      };
+    }
+
+    const bool: any = {
+      filter: [],
+    };
+    const subBool: any = {
+      should: [],
+    };
+    if (!!user) {
+      subBool.should.push({
+        term: {
+          ['permissions.' + permissionType + '.users']: user,
+        },
+      });
+      subBool.should.push({
+        term: {
+          ['permissions.' + permissionType + '.users']: '*',
+        },
+      });
+    }
+    if (!!group) {
+      subBool.should.push({
+        term: {
+          ['permissions.' + permissionType + '.groups']: group,
+        },
+      });
+      subBool.should.push({
+        term: {
+          ['permissions.' + permissionType + '.groups']: '*',
+        },
+      });
+    }
+
+    bool.filter.push({
+      bool: subBool,
+    });
+
+    if (!!savedObjectType) {
+      bool.filter.push({
+        terms: {
+          type: Array.isArray(savedObjectType) ? savedObjectType : [savedObjectType],
+        },
+      });
+    }
+
+    return { query: { bool } };
+  }
+
+  /* 
+    transfrom permissions format
+    input:   {
+        read: {
+            users:['user1']
+        },
+        write:{
+            groups:['group1']
+        }
+    }
+
+    output: [
+        {type:'user',name:'user1',permissions:['read']},
+        {type:'group',name:'group1',permissions:['write']},
+    ]
+    */
+  public transformPermissions(permissions: Permissions) {
```

Comment:
```suggestion
  public transformPermissions(permissions: Permissions): ACLTranformedResult {
```
It would be better that we export the type declare of the transformed result.

Reply:
Sure, I'll change it.